SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,436 @@
+      return baseAddress is AllocBoxInst
+    case .Stack:
+      return true
+    default:
+      return false
+    }
+  }
+  
+  /// Returns `true` if we can reliably compare for equality this `AccessBase`
+  /// with another `AccessBase`
+  var isUniquelyIdentified: Bool {
+    switch baseKind {
+    case .Box:
+      return baseAddress is AllocBoxInst
+    case .Stack, .Global:
+      return true
+    case .Argument:
+      return (baseAddress as! FunctionArgument).isExclusiveIndirectParameter
+    case .Class, .Tail, .Yield, .Pointer:
+      return false
+    }
+  }
+  
+  func isDistinct(from other: AccessBase) -> Bool {
+    if self.isUniquelyIdentified {
+      if other.isUniquelyIdentified {
+        if baseKind != other.baseKind { return true }
+        switch baseKind {
+        case .Global:
+          return (baseAddress as! GlobalAddrInst).global != (other.baseAddress as! GlobalAddrInst).global
+        case .Class: 
+          return (baseAddress as! RefElementAddrInst).fieldIndex != (other.baseAddress as! RefElementAddrInst).fieldIndex
+        case .Box, .Stack, .Argument, .Tail, .Yield, .Pointer:
+          return false
+        }
+      }
+    }
+    if other.isUniquelyIdentified { return other.isDistinct(from: self) }
+
+    if isObjectAccess {
+      if other.isObjectAccess {
+        if baseKind != other.baseKind { return true }
+
+        return baseKind == .Class &&
+               (baseAddress as! RefElementAddrInst).fieldIndex != (other.baseAddress as! RefElementAddrInst).fieldIndex
+      }
+
+      return false
+    }
+
+    if other.isObjectAccess {
+      return other.isDistinct(from: self)
+    }
+
+    return false
+  }
+}
+
+/// An `AccessPath` is a pair of a `base: AccessBase` and a `projectionPath: Path`
+/// which denotes the offset of the access from the base in terms of projections.
+struct AccessPath : CustomStringConvertible {
+  var base: AccessBase
+
+  /// address projections only
+  var projectionPath: SmallProjectionPath
+  
+  var description: String {
+    "\(projectionPath): \(base)"
+  }
+  
+  func isDistinct(from other: AccessPath) -> Bool {
+    if base.isDistinct(from: other.base) { return true }
+    return false
+  }
+}
+
+/// An `AccessStoragePath` is the reference (or a value for which the
+/// reference is a stored property) an address originates from.
+struct AccessStoragePath {
+  let storage: Value
+  
+  /// Only valid paths are: `"<sequence of value projections>.<one reference projection>.<sequence of address projections>"`
+  let path: SmallProjectionPath
+}
+
+typealias AccessStorage = (maybeRoot: AccessStoragePath?, base: AccessBase)
+
+enum EnclosingScope {
+  case scope(BeginAccessInst)
+  case base(AccessBase)
+}
+
+
+func canBeOperandOfIndexAddr(_ value: Value) -> Bool {
+  switch value {
+  case is IndexAddrInst, is RefTailAddrInst:
+    return true
+  default:
+    return false
+  }
+}
+
+fileprivate struct PointerIdentificationUseDefWalkerInternal : ValueUseDefWalker {
```

Comment:
nitpick: `Internal` doesn't need to be part of the name. It's already "documented" by the `fileprivate`.

Please structure the entries in the file top down, starting with the public API, then following the internally used utilities. So here: first `PointerIdentification`, then `PointerIdentificationUseDefWalker`.
In this case it probably makes sense to move this walker struct _into_ `PointerIdentification`.